[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69335
* #68770
* #68769
* __->__ #68229

Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

jerryzh168 committed:
[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op

Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

vkuzo commented:
is this only affecting the reference flow?  Do we plan to unfork subgraph rewriter before or after we move all customers to reference flow?  If this is going to be the production path for all customers, does it still make sense to fork?

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

vkuzo commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -72,8 +72,11 @@ def _lower_to_native_backend(model: QuantizedGraphModule) -> QuantizedGraphModul
-    model.recompile()
     for pattern, replacement in get_fbgemm_patterns_and_replacements():
-        subgraph_rewriter.replace_pattern(model, pattern, replacement)
+        while True:
```
should this be an iteration guard, to prevent an infinite loop if a bug sneaks in to `subgraph_rewriter.replace_pattern`?  Something like

```
# eliminate the possibility of an infinite loop
MAX_ITER = 1000
cur_iter = 0
while cur_iter < MAX_ITER:
  ...
  cur_iter += 1
```

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 commented on torch/ao/quantization/fx/quantized_fusion_patterns_and_replacements.py:
```diff
@@ -16,11 +18,207 @@ def relu_replacement(x, scale, zero_point):
+        def __init__(self):
+            super().__init__()
+            self.relu = torch.nn.ReLU()
+            self.scale = torch.tensor([])
+            self.zero_point = torch.tensor([])
```
if `self.scale` and `self.zero_point` are present in the original model we won't replace it: https://github.com/pytorch/pytorch/blob/master/torch/fx/subgraph_rewriter.py#L109

jerryzh168 commented on torch/ao/quantization/fx/quantized_fusion_patterns_and_replacements.py:
```diff
@@ -16,11 +18,207 @@ def relu_replacement(x, scale, zero_point):
+            x = qboprelu(x, num)
+            return x
+
+    return copy.deepcopy([
```
good point, this is added previously to make sure every time we generate a new object, but I'm not sure if this is still needed or not, let me check again

jerryzh168 commented on torch/ao/quantization/fx/subgraph_rewriter_FORKED_DO_NOT_USE.py:
```diff
@@ -49,13 +49,27 @@ def _match_nodes(self, pn: Node, gn: Node) -> bool:
         if pn in self.nodes_map:
             return self.nodes_map[pn] == gn
 
+        PRIM_TYPES = (int, float, torch.dtype)
```
yeah sure

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[wip][quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

vkuzo reviewed (approved):
quantization code looks good, I didn't review the subgraph rewriter code

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).

t10-13rocket reviewed (approved):
LGTM

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we rely on
subgraph_rewriter to rewrite the reference qunatized patterns to quantized ops

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537714](https://our.internmc.facebook.com/intern/diff/D32537714)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537714).
